File: java-workshop/part-3-mocks-clean/src/main/java/service/GatewayPaymentService.java
package service;

import domain.PaymentRequest;
import domain.PaymentResult;
import domain.PaymentStatus;
import gateway.PaymentGateway;

public class GatewayPaymentService implements PaymentService {
    private final PaymentGateway paymentGateway;

    public GatewayPaymentService(PaymentGateway paymentGateway) {
        this.paymentGateway = paymentGateway;
    }

    @Override
    public PaymentResult processPayment(String orderId, String customerId, double amount) {
        if (amount <= 0) {
            return new PaymentResult(orderId, PaymentStatus.FAILED, null);
        }

        PaymentRequest request = new PaymentRequest(orderId, customerId, amount);
        PaymentResult gatewayResult = paymentGateway.process(request);
        return new PaymentResult(orderId, gatewayResult.status(), gatewayResult.transactionId());
    }

    @Override
    public boolean refundPayment(String transactionId) {
        return paymentGateway.refund(transactionId);
    }
}

File: csharp-workshop/Part2.LightMocks/Service/GatewayPaymentService.cs
using Part2.LightMocks.Domain;
using Part2.LightMocks.Gateway;

namespace Part2.LightMocks.Service;

public class GatewayPaymentService : IPaymentService
{
    private readonly IPaymentGateway _paymentGateway;

    public GatewayPaymentService(IPaymentGateway paymentGateway)
    {
        _paymentGateway = paymentGateway;
    }

    public PaymentResult ProcessPayment(string orderId, string customerId, double amount)
    {
        if (amount <= 0)
        {
            return new PaymentResult(orderId, PaymentStatus.Failed, null);
        }

        var request = new PaymentRequest(orderId, customerId, amount);
        var gatewayResult = _paymentGateway.Process(request);
        return new PaymentResult(orderId, gatewayResult.Status, gatewayResult.TransactionId);
    }

    public bool RefundPayment(string transactionId)
    {
        return _paymentGateway.Refund(transactionId);
    }
}
